[DAGCombiner] Freeze maybe poison operands when folding select to logic

bjope opened the pull request:
Just like for regular IR we need to treat SELECT as conditionally
blocking poison in SelectionDAG. So (unless the condition itself is
poison) the result is only poison if the selected true/false value is
poison.
Thus, when doing DAG combines that turn SELECT into arithmetic/logical
operations (e.g. AND/OR) we need to make sure that the new operations
aren't more poisonous. One way to do that is to use FREEZE to make
sure the operands aren't posion.

This patch aims at fixing the kind of miscompiles reported in
  https://github.com/llvm/llvm-project/issues/84653
and
  https://github.com/llvm/llvm-project/issues/85190

Solution is to make sure that we insert FREEZE, if needed to make
the fold sound, when using the foldBoolSelectToLogic and
foldVSelectToSignBitSplatMask DAG combines.

bjope commented on llvm/test/CodeGen/AArch64/cmp-chains.ll:
```diff
@@ -6,21 +6,14 @@
 
 ; (x0 < x1) && (x2 > x3)
 define i32 @cmp_and2(i32 %0, i32 %1, i32 %2, i32 %3) {
```
I guess the typical problem here is that the arguments aren't marked as `noundef`. But even if they had been marked as `noundef` the lowering of arguments wouldn't resulting adding FREEZE nodes to indicate that the inputs aren't poison. Nor do we have a way to put such attributes on the CopyFromReg (or loads from stack) instructions that would map to accessing the arguments.

(No idea if there  is something in AArch64 that could be changed here to still optimize this in a valid way or if the old transform is unsafe from a poison/undef perspective.)

nikic replied:
Does SDAG ever look through CopyFromReg in some way? If not, we could assume that the value is frozen.

We can definitely treat function arguments as frozen in SDAG, I'm just not familiar with where else CopyFromReg gets used...

jayfoad replied:
> I'm just not familiar with where else CopyFromReg gets used

SDAG operates per basic block, and I believe CopyFromReg/CopyToReg are generated for any use/def of a "global" value, i.e. one that is live outside of the current basic block.

bjope commented on llvm/test/CodeGen/AArch64/fast-isel-select.ll:
```diff
@@ -295,22 +295,28 @@ define float @select_icmp_sle(i32 %x, i32 %y, float %a, float %b) {
 ; Test peephole optimizations for select.
 define zeroext i1 @select_opt1(i1 zeroext %c, i1 zeroext %a) {
 ; CHECK-LABEL: select_opt1
```
I figure that these diffs are a bit annoying. But given that the AssertZext/AssertSext are handles as maybe producing poison we can't hoist the freeze above it. And then I think we lose the optimizations based on that the inputs are masked down to a single bit. But maybe this also kind of originates from CopyFromReg not being treated as an implicit FREEZE (at least if the arguments would have been marked as `noundef`).

bjope replied:
Another thing here is that the fast-isel code still optimizes this case. So either there is a bug in fast-isel that it doesn't consider poison here. Or we should be allowed to optimize this also for the regular SelectionDAG ISel.
Worth mentioning is that GlobalISel isn't part of the checks here, but it seems like GlobalISel also would emit the explicit AND.

dtcxzyw commented:
Hi @bjope, could you please add the following test for #85190?
```
define i1 @pr85190(i64 %a) {
  %or = or i64 %a, 7
  %cmp1 = icmp slt i64 %a, 0
  %sub = sub nsw i64 -9223372036854775808, %or
  %cmp2 = icmp sgt i64 %sub, %a
  %res = select i1 %cmp1, i1 %cmp2, i1 false
  ret i1 %res
}
```

bjope commented:
> Hi @bjope, could you please add the following test for #85190?

Yes, thanks!
And that reminds me that I haven't added the reproducer for #84653 either yet.
I'll prepare a pre-commit with those tests.

dtcxzyw commented:
Ping for review :)

chenzheng1030 commented on llvm/test/CodeGen/PowerPC/pr40922.ll:
```diff
@@ -23,11 +23,12 @@ define i32 @a() {
-; CHECK-NEXT:    crorc 20, 1, 4
+; CHECK-NEXT:    andi. 5, 5, 1
+; CHECK-NEXT:    crnot 20, 4
+; CHECK-NEXT:    cror 20, 1, 20
```
This becomes worse. A combine for i1 type seems broken. Let me check why this happens.

chenzheng1030 replied:
```
  t43: i32,ch = load<(dereferenceable load (s32) from @a.b)> t10, t63, undef:i32
  t44: i32,glue = addc t43, Constant:i32<6>
            t46: i32 = and t44, Constant:i32<-17>
          t48: i1 = setcc t46, t43, setuge:ch
        t54: i1 = freeze t48    ;;This freeze breaks the instruction selection pattern for or(setcc) in PPC td files.
      t55: i1 = or t58, t54
```

bjope replied:
I think this is a situation when we do not fold the freeze over an operation (in this case a setcc) due to the operation having more then one operand that may be poison.
However, we got
```
    t43: i32,ch = load<(dereferenceable load (s32) from @a.b)> t10, GlobalAddress:i32<ptr @a.b> 0, undef:i32
    t44: i32,glue = addc t43, Constant:i32<6>
    t46: i32 = and t44, Constant:i32<-17>
    t48: i1 = setcc t46, t43, setuge:ch
    t54: i1 = freeze t48
```
so I think that t46 and t43 only could be poison/undef if t43 is poison/undef. So it would be enough to freeze t43, like this:
```
    t43: i32,ch = load<(dereferenceable load (s32) from @a.b)> t10, GlobalAddress:i32<ptr @a.b> 0, undef:i32
    t99: i32 = freeze t43
    t44: i32,glue = addc t99, Constant:i32<6>
    t46: i32 = and t44, Constant:i32<-17>
    t54: i1 = setcc t46, t99, setuge:ch
```

Not exactly sure, but maybe if it would be possible to improve the analysis in DAGCombiner::visitFREEZE. Idea would be to allow pushing freeze through more operations (with multiple maybe poison/undef operands as we do for BUILD_VECTOR etc)). One could limit it to the scenraio when the isGuaranteedNotToBeUndefOrPoison checks for those operands reaches the same DAG node when recursing, i.e. when the likelyhood is that we still would end up with a single freeze in the end.

nikic commented:
> Ping for review :)

The regressions needs to be mitigated to the maximum extent feasible before we make this change.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -11307,28 +11307,34 @@ static SDValue foldBoolSelectToLogic(SDNode *N, SelectionDAG &DAG) {
-  // select Cond, Cond, F --> or Cond, F
-  // select Cond, 1, F    --> or Cond, F
+  auto FreezeIfNeeded = [&](SDValue V) {
+    if (!DAG.isGuaranteedNotToBePoison(V))
```
getNode(ISD::FREEZE) should do this for us?
https://github.com/llvm/llvm-project/blob/02cb89b36a7ae9be4ab657306b69dc9d2830d0d5/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L5718-L5723

bjope replied:
Ah. Nice!
Downstream I also needed some additional checks to just do this for my target (since it is a pain doing it for all targets until this has landed upstream). So that is one reason why I've wrapped the check this way.

AtariDreams commented:
Ping?

RKSimon commented on llvm/test/CodeGen/X86/vector-compare-all_of.ll:
```diff
@@ -1543,34 +1543,49 @@ define i1 @select_v2i8(ptr %s0, ptr %s1) {
+; SSE2-NEXT:    movl %ecx, %eax
+; SSE2-NEXT:    andb $2, %al
+; SSE2-NEXT:    shrb %al
+; SSE2-NEXT:    andb %cl, %al
```
I'm not certain but I think these regressions (and vector-compare_any_of.ll) will go away with #87480

RKSimon replied:
@bjope  #87480 has landed if you want to rebase and see if we have any improvements

arsenm commented on llvm/test/CodeGen/AArch64/fast-isel-select.ll:
```diff
@@ -295,22 +295,28 @@ define float @select_icmp_sle(i32 %x, i32 %y, float %a, float %b) {
-; CHECK:       orr {{w[0-9]+}}, [[REG]], w1
+; SISEL:       orn [[REG:w[0-9]+]], w1, w0
+; SISEL:       and w0, [[REG]], #0x1
+; FISEL:       eor [[REG:w[0-9]+]], w0, #0x1
+; FISEL:       orr {{w[0-9]+}}, [[REG]], w1
   %1 = select i1 %c, i1 %a, i1 true
   ret i1 %1
 }
 
 define zeroext i1 @select_opt3(i1 zeroext %c, i1 zeroext %a) {
 ; CHECK-LABEL: select_opt3
-; CHECK:       bic {{w[0-9]+}}, w1, w0
```
Should we add copies of regression testcases with noundef arguments that show the original output?

bjope replied:
Adding noundef to the arguments doesn't help here. There will still be an AssertZext being added when lowering the arguments.
Not quite sure how we should deal with the (freeze(AssertZext)).

arsenm replied:
I guess we should have an equivalent AssertNoundef or something, but the intersection of all the Assert*s corresponding to argument attributes is not great

bjope replied:
Well, one idea is that we get rid of the known, IMHO, misuse of AssertSext/AssertZext in DAGTypeLegalizer::PromoteIntRes_FP_TO_XINT. And then try to check all other places that create AssertSext/AssertZext nodes to see if they are violating the desciption of those ISD nodes (i.e. that they "record if a register contains a value that has already been zero or sign extended"). In the PromoteIntRes_FP_TO_XINT those nodes are used to indicate that it is the result is poison if the input violates the assert. But the desciption of the assert nodes just say that the result is poison if the input is poison.
This probably result in various regressions, so maybe not acceptable.

Another idea would be to add some kind of boolean parameter to AssertSext/AssertZext that makes it possible to indicate that the assert node only propagates posion or if it can create poison. This flag could then be used in situations like PromoteIntRes_FP_TO_XINT.

OTOH, considering that https://github.com/llvm/llvm-project/issues/66603 still is open. Maybe the revert in https://github.com/llvm/llvm-project/commit/8505c3b15bfc535ff6624e71add4082680745187 wasn't needed after all?

nikic replied:
> OTOH, considering that #66603 still is open. Maybe the revert in [8505c3b](https://github.com/llvm/llvm-project/commit/8505c3b15bfc535ff6624e71add4082680745187) wasn't needed after all?

The issue is still open due to an optimization problem -- the revert was needed to fix a correctness problem.

> Another idea would be to add some kind of boolean parameter to AssertSext/AssertZext that makes it possible to indicate that the assert node only propagates posion or if it can create poison. This flag could then be used in situations like PromoteIntRes_FP_TO_XINT.

I think to make this actually useful the statement would have to be that violating the AssertZext with the flag is immediate UB (including the case where the input is poison).

I do think that something in this direction would make sense.

Something I'm unsure about is to what degree nodes with IUB in SDAG are safe without a chain operand -- we have known miscompiles from the fact that division by zero UB is not modeled via chain.

bjope commented:
I've rebased this (again). But I'm not quite sure how to proceed.
- What remains to be solved before landing this? Any particular tests that are of concern?
- Would it help if I split it up in one patch for foldBoolSelectToLogic and another one for foldVSelectToSignBitSplatMask?
- Should I pre-commit https://github.com/llvm/llvm-project/pull/84924/commits/e57c1432d57227969a9d6a4f19fd5469a39a4a14 (not sure if there would be any motivating test cases for that one if doing it as a pre-commit, that is why I added it as a second patch in the patch-stack here).

nikic commented:
> I've rebased this (again). But I'm not quite sure how to proceed.
> 
>     * What remains to be solved before landing this? Any particular tests that are of concern?

I think we mainly need a solution for the "AssertZExt/AssertSExt on argument" problem.

> 
>     * Would it help if I split it up in one patch for foldBoolSelectToLogic and another one for foldVSelectToSignBitSplatMask?

Probably, especially if the latter has less test regressions?
 
>     * Should I pre-commit [e57c143](https://github.com/llvm/llvm-project/commit/e57c1432d57227969a9d6a4f19fd5469a39a4a14) (not sure if there would be any motivating test cases for that one if doing it as a pre-commit, that is why I added it as a second patch in the patch-stack here).

Yes, please. If there are no test changes as-is, you could add a test with explicit freeze.

bjope commented:
> > ```
> > * Would it help if I split it up in one patch for foldBoolSelectToLogic and another one for foldVSelectToSignBitSplatMask?
> > ```
> 
> Probably, especially if the latter has less test regressions?
> 
> > ```
> > * Should I pre-commit [e57c143](https://github.com/llvm/llvm-project/commit/e57c1432d57227969a9d6a4f19fd5469a39a4a14) (not sure if there would be any motivating test cases for that one if doing it as a pre-commit, that is why I added it as a second patch in the patch-stack here).
> > ```
> 
> Yes, please. If there are no test changes as-is, you could add a test with explicit freeze.

Thanks for the advice! I'll split this patch in smaller pieces (at least breaking out the "Push freeze through SETCC and SELECT_CC" part).

efriedma-quic commented:
Maybe we should introduce ISD::LOGICAL_AND/ISD::LOGICAL_OR nodes to SelectionDAG?  If we delay the lowering to actual AND/OR until operation legalization, we can do various combines without worrying about a bunch of extra freeze operations.

nikic commented:
> Maybe we should introduce ISD::LOGICAL_AND/ISD::LOGICAL_OR nodes to SelectionDAG? If we delay the lowering to actual AND/OR until operation legalization, we can do various combines without worrying about a bunch of extra freeze operations.

Yes, that would be an alternative -- this is essentially what we do in IR, though there we just keep the original select but recognize it as a logical and/or pattern. The advantage is that we don't have to introduce freeze -- the disadvantage is that we have to make sure that all optimizations can handle the logical variant. On the IR side, this took some very extensive work. Maybe it's simpler for SDAG as it presumably has less transforms reasoning about complex conditions.

dtcxzyw commented:
Reverse ping :)
I encountered this issue in real-world code (generated by verilator for RTL simulation).

nikic commented:
With the LLVM 19 release coming up, we might have to take this as-is :(

AtariDreams commented:
Any update on this before LLVM-19 ?

nikic commented:
@bjope Can you please rebase this PR?

bjope commented:
Seems like I never splitted this PR into two before going on vacation. I'll fix that and rebase.

bjope commented on llvm/test/CodeGen/X86/combine-pmadd.ll:
```diff
@@ -229,10 +229,41 @@ define i32 @combine_pmaddubsw_constant_sat() {
-; CHECK:       # %bb.0:
-; CHECK-NEXT:    movb $1, %al
-; CHECK-NEXT:    retq
+; SSE-LABEL: pmaddwd_pcmpgt_infinite_loop:
```
Looks like we get:
```
   t3: v8i16 = BUILD_VECTOR Constant:i16<-32768>, Constant:i16<-32768>, Constant:i16<-32768>, Constant:i16<-32768>, Constant:i16<-32768>, Constant:i16<-32768>, Constant:i16<-32768>, Constant:i16<-32768>
   t4: v4i32 = llvm.x86.sse2.pmadd.wd TargetConstant:i64<13393>, t3, t3
   t62: v4i32 = freeze t4
```

X86 can perhaps model that intrinsic as not creating poison. But a more general solution could be to let SelectionDAG::computeKnownBits look through FREEZE (and then I guess the same should be done for SelectionDAG::ComputeNumSignBits as well).

bjope replied:
Btw, it is a bit weird that SelectionDAG::computeKnownBits is handling AssertZext as if the upper bits are known to be zero. At least as long as we say that AssertZext may result in a poison value that is violating the asserted condition.

But if we consider these three examples, then I think computeKnownBits on %3 should give the same result, right?
```
  %1 = AssertZext i32 %0, i1
  %2 = lshr i32 %1, 1
  %3 = freeze i32 %2
```  
```
  %1 = AssertZext i32 %0, i1
  %2 = freeze i32 %1
  %3 = lshr i32 %2, 1
``` 
```
  %1 = AssertZext i32 %0, i1
  %3 = lshr i32 %1, 1
```

nikic replied:
All computeKnownBits() style APIs return "known or poison" results. It's not possible for look through freeze.

bjope replied:
> All computeKnownBits() style APIs return "known or poison" results. It's not possible for look through freeze.

Right. Hm. And pushing freeze through operations aggressively/early might block other simplifications as in the example with lshr (which is why we currently avoid it for SRA and SRL).

Another idea then is to let DAGCombiner::visitFREEZE (or getNode for FREEZE) constant fold the operand (checking if computeKnownBits returns a constant). This to avoid that we mess up constant folding by pushing the freeze through an operation that hasn't been constant folded yet.
Or is that also incorrect given that computeKnownBits has returned "known or poison"? If any constant folded value that has been derived via computeKnownBits is a possible poison value, then how can SelectionDAG::isGuaranteedNotToBeUndefOrPoison return true for constants (we can't know if a constant is a possible posion value.

bjope replied:
Anyway, maybe we can rely on that for example intrinsic calls like the one here has been constant folded already before ISel. And this test case isn't important to optimize (it is just a regression test for something that used to hit an infinite loop)..

bjope replied:
> Another idea then is to let DAGCombiner::visitFREEZE (or getNode for FREEZE) constant fold the operand (checking if computeKnownBits returns a constant). This to avoid that we mess up constant folding by pushing the freeze through an operation that hasn't been constant folded yet. Or is that also incorrect given that computeKnownBits has returned "known or poison"? If any constant folded value that has been derived via computeKnownBits is a possible poison value, then how can SelectionDAG::isGuaranteedNotToBeUndefOrPoison return true for constants (we can't know if a constant is a possible posion value.

To answer that myself, constant folding the freeze should be ok. But just looking through FREEZE in computeKnownBits etc is not valid.

RKSimon replied:
I'll get X86ISD VPMADDUBSW/VPMADDWD handling added to canCreateUndefOrPoisonForTargetNode shortly

RKSimon replied:
e9974f02a30280eab868cdab0a90a5372b1799cd

bjope commented:
> With the LLVM 19 release coming up, we might have to take this as-is :(

So can I land this now? Branch out for llvm 19 is tomorrow.

nikic commented on llvm/test/CodeGen/AArch64/fast-isel-select.ll:
Please rebase over https://github.com/llvm/llvm-project/commit/a7fb25dd1fcc2e5afcc65cccfa83b7b381b48906. Hard to understand what actually changed here.

bjope committed:
[DAGCombiner] Freeze maybe poison operands when folding select to logic

Just like for regular IR we need to treat SELECT as conditionally
blocking poison. So (unless the condition itself is poison) the
result is only poison if the selected true/false value is poison.
Thus, when doing DAG combines that turn SELECT into arithmetic/logical
operations (e.g. AND/OR) we need to make sure that the new operations
aren't more poisonous. One way to do that is to use FREEZE to make
sure the operands aren't posion.

This patch aims at fixing the kind of miscompiles reported in
  https://github.com/llvm/llvm-project/issues/84653
and
  https://github.com/llvm/llvm-project/issues/85190

Solution is to make sure that we insert FREEZE, if needed to make
the fold sound, when using the foldBoolSelectToLogic and
foldVSelectToSignBitSplatMask DAG combines.

nikic commented:
> > With the LLVM 19 release coming up, we might have to take this as-is :(
> 
> So can I land this now? Branch out for llvm 19 is tomorrow.

Could you please land https://github.com/llvm/llvm-project/pull/94492 first and then rebase?

bjope commented:
@nikic : Rebased (now on top of https://github.com/llvm/llvm-project/pull/94492, and https://github.com/llvm/llvm-project/commit/a7fb25dd1fcc2e5afcc65cccfa83b7b381b48906).

@RKSimon : Also on top of https://github.com/llvm/llvm-project/commit/e9974f02a30280eab868cdab0a90a5372b1799cd , but I don't think that one made any difference here.

nikic reviewed (approved):
LGTM

bjope commented:
> LGTM

Thanks!
